windows/devices/uwb/UwbConnector.cxx

```diff
@@ -641,111 +692,90 @@ UwbConnector::HandleNotifications(std::stop_token stopToken)
     LOG_INFO << "uwb notification listener stopped for device " << DeviceName();
 }
 
-std::vector<std::shared_ptr<::uwb::UwbRegisteredDeviceEventCallbacks>>
-UwbConnector::GetResolvedDeviceEventCallbacks()
+/**
+ * @brief Internal helper function to invoke the callbacks with the arg
+ * If the weak_ptr callback held within a token is not valid, then the token is removed
+ *
+ * @tparam ArgTs the type of the args to be passed into the callbacks
+ * @tparam TokenT the type of the token
+ * @param tokens
+ * @param args
+ */
+template <typename TokenT, typename... ArgTs>
+void
+InvokeCallbacks(std::vector<std::shared_ptr<TokenT>>& tokens, ArgTs&... args)
 {
-    std::lock_guard eventCallbacksLockExclusive{ m_eventCallbacksGate };
-
-    std::vector<std::shared_ptr<::uwb::UwbRegisteredDeviceEventCallbacks>> deviceEventCallbacks;
-    deviceEventCallbacks.reserve(std::size(m_deviceEventCallbacks));
+    if (tokens.empty()) {
+        PLOG_INFO << "Ignoring " << typeid(TokenT).name() << " event due to missing callbacks";
+    }
 
-    for (auto it = std::begin(m_deviceEventCallbacks); it != std::end(m_deviceEventCallbacks);) {
-        auto& deviceEventCallbackToken = *it;
-        auto& deviceEventCallbackWeak = deviceEventCallbackToken->Callbacks;
-        auto deviceEventCallback = deviceEventCallbackWeak.lock();
-        if (deviceEventCallback != nullptr) {
-            deviceEventCallbacks.push_back(std::move(deviceEventCallback));
-            it = std::next(it);
+    for (auto it = std::cbegin(tokens); it != std::cend(tokens);) {
+        auto token = *it;
+        auto callbackWeak = token->Callback;
+        auto callbackShared = callbackWeak.lock();
+        if (not callbackShared) {
+            it = tokens.erase(it);
         } else {
-            it = m_deviceEventCallbacks.erase(it);
+            auto callback = *callbackShared;
+            if (not callback) {
+                it = tokens.erase(it);
+            }
+            callback(args...);
+            it = std::next(it);

```

Comment:
Should invocation of the callback and advancing the iterator be in an `else` clause? Otherwise, it seems `callback` is de-referenced when it's null and `it` is advanced twice (once with `erase`, another with `std::next(it)`). 

Also, I believe `args` should be forwarded to optimize out copies (this necessitates making `args` a universal reference, `ArgTs&&... args`):

```suggestion
            if (not callback) {
                it = tokens.erase(it);
            } else {
                callback(std::forward<ArgTs>(args));
                it = std::next(it);
            }
```

Reply:
good suggestion, updated